src/main/java/school/lemon/changerequest/java/collections/hw1/ExtendedListImpl.java

```diff
@@ -0,0 +1,709 @@
+        public List<E> subList(int fromIndex, int toIndex) {
+            checkSubListIndexes(fromIndex, toIndex);
+            checkConcurrentModification();
+            return new SubList(this, begin, fromIndex, toIndex);
+        }
+
+        @Override
+        public String toString() {
+            StringBuilder result = new StringBuilder("[");
+            if (size > 0)
+                result.append(ExtendedListImpl.this.container[begin]);
+            for (int i = 1; i < size; i++)
+                result.append(", ").append(ExtendedListImpl.this.container[begin + i]);
+            result.append("]");
+            return result.toString();
+        }
+
+        private void isCorrectIndex(int index) {
+            if (index < 0 || index >= size)
+                throw new IndexOutOfBoundsException("Incorrect index:" + index);
+        }
+
+        private void isCorrectIndexForAdd(int index) {
+            if (index < 0 || index > size)
+                throw new IndexOutOfBoundsException("Incorrect index:" + index);
+        }
+
+        private void checkSubListIndexes(int fromIndex, int toIndex) {
+            if (fromIndex < 0 || toIndex > size || fromIndex > toIndex)
+                throw new IndexOutOfBoundsException("Incorrect index:");
+        }
+
+        private void checkConcurrentModification() {
+            if (modificationCoef != ExtendedListImpl.this.modificationCoef)
+                throw new ConcurrentModificationException();
+        }
+
+    }
+
+
+    /**********************************************************************************/
+
+    private class ListItr extends Itr implements ListIterator<E> {
+
+        ListItr(List<E> iteratedList) {
+            this(iteratedList, 0);
+        }
+
+        ListItr(List<E> iteratedList, int index) {
+            super(iteratedList);
+            cursor = index;
+        }
+
+        @Override
+        public boolean hasPrevious() {
+            return cursor > 0;
+        }
+
+        @Override
+        public E previous() {
+            if (!hasPrevious())
+                throw new NoSuchElementException("List doesn`t contain more elements");
+            lastPosition = --cursor;
+            return iteratedList.get(cursor);
+        }
+
+        @Override
+        public int nextIndex() {
+            return cursor;
+        }
+
+        @Override
+        public int previousIndex() {
+            return cursor - 1;
+        }
+
+        @Override
+        public void set(E e) {
+            if (lastPosition < 0)
+                throw new IllegalStateException("Method next or previous was not called");
+            iteratedList.set(lastPosition, e);
+        }
+
+        @Override
+        public void add(E e) {
+            iteratedList.add(cursor, e);
+            cursor++;
+            lastPosition = -1;
+        }
+    }
+
+    /**********************************************************************************/
+
+    private class Itr implements Iterator<E> {
```

Comment:
Try to avoid such names, while they do not provide any suitable context within the name.

Probably, something like `DirectIterator` or just `IteratorImpl` would be better here.